[improve][broker] Pulsar Rate Limiting Refactoring changes (PIP-322)

lhotari opened the pull request:
Fixes #21442
PIP: #21680 

### Motivation

See [PIP-322](https://github.com/apache/pulsar/blob/master/pip/pip-322.md) for details.

### Modifications

See [PIP-322](https://github.com/apache/pulsar/blob/master/pip/pip-322.md) for details.

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

lhotari commented:
/pulsarbot rerun-failure-checks

onobc commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java:
```diff
@@ -852,4 +852,25 @@ public boolean isDisconnecting() {
     private static final Logger log = LoggerFactory.getLogger(Producer.class);
 
+    /**
+     * Increments the counter that controls the throttling of the connection by pausing reads.
```
**[Nit]** This sounds like it increments the counter by pausing reads but really it just increments the counter and the counter is described as... (same for decrement method). I think something like this may clarify it a bit, wdyt?

```
    /**
     * Increments the counter that controls the throttling of the connection.
     * <p>
     * When the counter is greater than 0, the connection is throttled by pausing reads.
     * <p>
     * The caller is responsible for decrementing the counter by calling {@link #decrementThrottleCount()}  when the
     * connection should no longer be throttled.
     */
```

lhotari replied:
I made improvements to the javadoc. PTAL

onobc commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PublishRateLimiterImpl.java:
```diff
@@ -16,70 +16,133 @@
-        return publishRateExceeded;
+    private void unthrottleQueuedProducers(ScheduledExecutorService executor) {
+        Producer producer;
+        // unthrottle producers until the token buckets contain tokens
```
**[Nit]** I think this should be  "unthrottle producers until the token buckets do not contain tokens" or "unthrottle as many producers as possible while there are token available".

lhotari replied:
Thanks. Applied.

onobc commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PublishRateLimiterImpl.java:
```diff
@@ -16,70 +16,133 @@
+            // consume tokens from the token bucket for bytes
+            currentTokenBucketOnByte.consumeTokens(msgSizeInBytes);
+            // check if the token bucket contains remaining tokens, if not, we should throttle
+            shouldThrottle = shouldThrottle || !currentTokenBucketOnByte.containsTokens();
```
**[Nit]** Could simplify a bit w/ `|=`

onobc commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PublishRateLimiterImpl.java:
```diff
@@ -16,70 +16,133 @@
+            // check if the token bucket contains remaining tokens, if not, we should throttle
+            shouldThrottle = shouldThrottle || !currentTokenBucketOnByte.containsTokens();
+        }
+        if (shouldThrottle) {
```
The producer has no current knowledge of why it was rate limited (# msgs or # bytes). Do you envision adding context in a later PR (post V1)?

lhotari replied:
yes, that's possible.

onobc commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:
```diff
@@ -898,57 +888,35 @@ public long increasePublishLimitedTimes() {
+        } else {
+            activeRateLimiters = List.of(this.topicPublishRateLimiter, getBrokerPublishRateLimiter());
         }
+        activeRateLimiters = activeRateLimiters.stream().filter(Objects::nonNull).toList();
```
I actually meant waiting to assign the new list to the instance var once the list is complete and ready. There is a rare chance that between line 908 and 914 that someone accesses the null limiter that is being filtered out on 914. 

Something like this: 
```java
   private void updateActiveRateLimiters() {
        List<PublishRateLimiter> updatedRateLimiters = new ArrayList<>();
        if (isResourceGroupRateLimitingEnabled()) {
            updatedRateLimiters =
                    List.of(this.topicPublishRateLimiter, getBrokerPublishRateLimiter(), resourceGroupPublishLimiter);
        } else {
            updatedRateLimiters = List.of(this.topicPublishRateLimiter, getBrokerPublishRateLimiter());
        }
        activeRateLimiters = updatedRateLimiters.stream().filter(Objects::nonNull).toList();
}
```

lhotari replied:
Thanks @onobc, very useful. Missed that. Fixing.

lhotari replied:
Fixed.

onobc reviewed (approved):
This looks good @lhotari . I have left a few minor comments. 

One concern is of course the size of the PR and the large feature this is replacing. I think this is a known though and we will rely on current tests and getting some runtime on this sooner rather than later to help w/ this risk. I do think we should beef up the tests in this area to cover the critical path and form a larger integration test perspective.

lhotari commented:
> This looks good @lhotari . I have left a few minor comments.
> 
> One concern is of course the size of the PR and the large feature this is replacing. I think this is a known though and we will rely on current tests and getting some runtime on this sooner rather than later to help w/ this risk. I do think we should beef up the tests in this area to cover the critical path and form a larger integration test perspective.

Thanks for the review @onobc! Very useful suggestions too. I'll follow up with the changes soon.

onobc reviewed (approved):
Thanks for the update @lhotari - looks great.

onobc commented:
Nice updates @lhotari. I also noticed the code coverage numbers are going ⬆️
 🎉

lhotari commented:
> Nice updates @lhotari. I also noticed the code coverage numbers are going ⬆️ 🎉

Thanks @onobc . btw. The code coverage metrics aren't reliable because of the flaky tests that cause the master branch builds to fail very often. The comparison is made against the latest master branch build and when some parts of the build fail, it won't contain a complete result. Code coverage is collected from about 25 different build jobs.

Shawyeok commented on pulsar-common/src/main/java/org/apache/pulsar/common/util/AsyncTokenBucket.java:
```diff
@@ -0,0 +1,483 @@
+ * 1. tokens are consumed by calling the "consumeTokens" method.
+ * 2. the "calculateThrottlingDuration" method is called to calculate the duration of a possible needed pause when the
+ * tokens
+ * are fully consumed.
```
- It would be more clear if tell user how to determine tokens are whether fully consumed.
- Is it possible to tell is there any tokens available after `consumeTokens` method exit.

lhotari replied:
> It would be clearer if we could inform the user about how to determine whether the tokens are fully consumed.

thanks for the feedback. I'll try to clarify this by making a further refactoring. 
In the current design, tokens are eventually consistent, so it would be useful to have this part of the naming.
For the asynchronous rate limiting use cases in Pulsar, there isn't really a need to have a consistent view of tokens. The token value can also be negative and that is important for correct behavior of the rate limiter.

lhotari replied:
@Shawyeok PTAL, I have clarified the Javadoc. I added a separate method consumeTokensAndCheckIfContainsTokens to handle the main usage flow in PublisherLimiterImpl. The docs contain explanations about the eventual consistent nature.

lhotari replied:
@Shawyeok PTAL. I have also updated the [PIP-322 document](https://github.com/lhotari/pulsar/blob/lh-ratelimiter-refactoring-pip/pip/pip-322.md) with better explanations.

merlimat reviewed (approved):
Great work here!

Can we also include the JMH benchmark from https://github.com/lhotari/async-tokenbucket/blob/master/src/jmh/java/com/github/lhotari/asynctokenbucket/AsyncTokenBucketBenchmark.java . It would be great to use it down the road for validating further changes to the implementation.

lhotari commented on pulsar-common/src/main/java/org/apache/pulsar/common/util/AsyncTokenBucket.java:
```diff
@@ -0,0 +1,615 @@
+
+        @Override
+        public void close() {
+            closed = true;
```
I'll switch to use `Thread.interrupt()`

lhotari replied:
Done

lhotari commented on pulsar-common/src/main/java/org/apache/pulsar/common/util/AsyncTokenBucket.java:
```diff
@@ -0,0 +1,615 @@
+    /**
+     * Interface for a clock source that returns a monotonic time in nanoseconds with a required precision.
+     */
+    public interface MonotonicClockSource {
```
+1. The reason to put everything in a single file was to make AsyncTokenBucket self contained at first. I'm thinking of moving all AsyncTokenBucket files to a separate "qos" package.

One detail about this MonotonicClockSource interface is designed for AsyncTokenBucket's use case, so I'll keep it package private for now so that the usages can be controlled whenever that happens.

lhotari replied:
Resolved by moving to separate file and renaming to MonotonicSnapshotClock

lhotari commented on pulsar-common/src/main/java/org/apache/pulsar/common/util/AsyncTokenBucket.java:
```diff
@@ -0,0 +1,615 @@
+         *                      if false, the returned value can be a granular precision monotonic time in nanoseconds.
+         * @return the current monotonic clock time in nanoseconds
+         */
+        long getNanos(boolean highPrecision);
```
This is needed for AsyncTokenBucket's use case. I'll try to make it more understandable. The "highPrecision" is now perhaps causing the confusion. The use case in AsyncTokenBucket is with the usages where a consistent view of the token balance is requested. 

I'll rename the parameter or add documentation to make it clear why there's both high/low precision in the same interface and method.

lhotari replied:
Update: I am renaming MonotonicClockSource to MonotonicSnapshotClock and making it explicit that it's a clock that snapshots the underlying high precision clock source periodically. I am renaming "highPrecision" to "requestSnapshot" to make it clear that when the flag is set, the snapshot is updated in the call before returning the value.

lhotari replied:
Done. PTAL.

lhotari commented on pulsar-common/src/main/java/org/apache/pulsar/common/util/AsyncTokenBucket.java:
```diff
@@ -0,0 +1,615 @@
+         *                      if false, the returned value can be a granular precision monotonic time in nanoseconds.
+         * @return the current monotonic clock time in nanoseconds
+         */
+        long getNanos(boolean highPrecision);
```
+1, I'll rename it. `getCurrentTickNanos` perhaps. The reason for this is that `time` could be confusing since it's not the time of the day (like it is in `System.currentTimeMillis()`. I think `tick` might be a better description of what the returned value is.

lhotari replied:
Renaming to `getTickNanos`

lhotari replied:
Done.

lhotari commented on pulsar-common/src/main/java/org/apache/pulsar/common/util/AsyncTokenBucket.java:
```diff
@@ -0,0 +1,615 @@
+     * Instantiating this class creates a daemon thread that updates the monotonic time. The close method
+     * should be called to stop the thread.
+     */
+    public static class GranularMonotonicClockSource implements MonotonicClockSource, AutoCloseable {
```
+1

lhotari replied:
Done.

lhotari commented on pulsar-common/src/main/java/org/apache/pulsar/common/util/AsyncTokenBucket.java:
```diff
@@ -0,0 +1,615 @@
+ * constructing higher-level asynchronous rate limiter implementations, which require side effects for throttling.
+ * <p>To achieve optimal performance, pass a {@link GranularMonotonicClockSource} as the clock source.
+ */
+public abstract class AsyncTokenBucket {
```
+1

lhotari replied:
Done.

lhotari commented:
> Great work here!
> 
> Can we also include the JMH benchmark from https://github.com/lhotari/async-tokenbucket/blob/master/src/jmh/java/com/github/lhotari/asynctokenbucket/AsyncTokenBucketBenchmark.java . It would be great to use it down the road for validating further changes to the implementation.

@merlimat  Thanks for that review! Great idea! I added a module "microbench". There's also a README.md file that explains how to compile and run the JMH benchmarks.
I have now addressed your review comments. PTAL

nodece commented:
This PR looks to break the resourcegroup rate limiter, please see the Prometheus graph:

<img width="1781" alt="image" src="https://github.com/apache/pulsar/assets/16235121/c8b9be89-02cf-4eb4-8267-3c8f94654884">

When using the 3.3.0 version, the `pulsar_rate_in` metric is not as smooth as the 3.0.1 version.

The following is the broker.conf:
```
PULSAR_PREFIX_resourceUsageTransportClassName=org.apache.pulsar.broker.resourcegroup.ResourceUsageTopicTransportManager
```

Test:
```
bin/pulsar-admin resourcegroups create rg-test-1 -mp 200
bin/pulsar-admin namespaces set-resource-group public/default -rgn rg-test-1
bin/pulsar-perf produce rg-test-1 --disable-batching -r 600 --size 10
```

Could you help to check out this issue?
